packages/run-protocol/src/vaultFactory/liquidateIncrementally.js

```diff
@@ -0,0 +1,322 @@
+  const nextBlock = async () => E(timerService).delay(1n);
+
+  // TODO make subject to governance
+  const SCALE = 1_000_000n;
+  const BASIS_POINTS = 10_000n * SCALE;
+  const BP2 = BASIS_POINTS * BASIS_POINTS;
+
+  const AMM_FEE_BP = 30n * SCALE;
+  const MAX_IMPACT_BP = 50n * SCALE;
+  // sqrt(impact + 1) - 1
+  // Since we take SQRT but want to be BP, multiply in BP
+  const IMPACT_FACTOR_BP =
+    BigInt(
+      Math.ceil(
+        Math.sqrt(Number((MAX_IMPACT_BP + BASIS_POINTS) * BASIS_POINTS)),
+      ),
+    ) - BASIS_POINTS;
+  const IMPACT_FACTOR_BP2 = IMPACT_FACTOR_BP * (BASIS_POINTS - AMM_FEE_BP);
+
+  const asFloat = (numerator, denominator) =>
+    Number(numerator) / Number(denominator);
+  trace(
+    'FACTORS',
+    asFloat(AMM_FEE_BP, BASIS_POINTS),
+    asFloat(MAX_IMPACT_BP, BASIS_POINTS),
+    Math.sqrt(Number(MAX_IMPACT_BP + BASIS_POINTS)),
+    asFloat(IMPACT_FACTOR_BP, BASIS_POINTS),
+    asFloat(IMPACT_FACTOR_BP2, BP2),
+  );
+
+  // How close the sale price on the AMM must be to the expected price from the oracle
+  // quote must be at least 70% 0f what the oracle says
+  const ORACLE_TOLERANCE = 3000n * SCALE;
+  const ORACLE_LIMIT_BP = makeRatio(
+    BASIS_POINTS - ORACLE_TOLERANCE,
+    debtBrand,
+    BASIS_POINTS,
+  );
+
+  // 3% max slippage = 97% min
+  const MAX_SLIPPED = makeRatio(9700n * SCALE, debtBrand, BASIS_POINTS);
+
+  trace('CONSTANTS', {
+    IMPACT_FACTOR_BP,
+    BASIS_POINTS,
+    IMPACT_FACTOR_BP2,
+    BP2,
+  });
+
+  // the maximum number fo failure sina  row before we stop liquidating.
+  // TODO make this larger andmake it reset on success
+  const MAX_FAILURES = 4;
+  let failures = 0;
+
+  const computeAMMProceeds = (poolCentral, poolCollateral, tranche, debt) => {
+    const k = NatMath.multiply(poolCentral.value, poolCollateral.value);
+    const postSaleCollateral = AmountMath.add(tranche, poolCollateral);
+    const quoteValue = NatMath.subtract(
+      poolCentral.value,
+      NatMath.floorDivide(k, postSaleCollateral.value),
+    );
+    const ammQuote = AmountMath.make(debt.brand, quoteValue);
+    const minAmmProceeds = ceilMultiplyBy(ammQuote, MAX_SLIPPED);
+    trace('AMM estimate', {
+      tranche,
+      minAmmProceeds,
+      poolCentral,
+      poolCollateral,
+    });
+    return minAmmProceeds;
+  };
+
+  /**
+   * Compute the tranche size whose sale on the AMM would have
+   * a price impact of MAX_IMPACT_BP.
+   * This doesn't use ratios so that it is usable for any brand
+   *
+   * @param {Amount} poolSize
+   * @returns {Amount}
+   */
+  const maxTrancheWithFees = poolSize =>
+    AmountMath.make(
+      poolSize.brand,
+      NatMath.ceilDivide(
+        NatMath.multiply(poolSize.value, IMPACT_FACTOR_BP2),
+        BP2,
+      ),
+    );
+
+  /**
+   *
+   * @param {ZCFSeat} seat
+   * @param {Amount} originalDebt
+   * @yields {LiquidationStep}
+   */
+  async function* processTranches(seat, originalDebt) {
```

Comment:
not a blocker but generators won't work in Jessie or TinySES.

Reply:
Correct. This is outside the scope of those.